### PR TITLE
Credential Library - Create

### PR DESCRIPTION
**Pull request by gsusmi**
Screen Shot 2021-07-18 at 14 30 51

### CODE REVIEW

File: addons/core/translations/form/en-us.yaml [hunk 1]
```diff
@@ -115,3 +115,12 @@ value:
...
   help: The URL to which the third-party provider sends the authentication response.  Paste this into the provider configuration.
+vault_path:
+  label: Vault Path
+  help: The path in Vault to request credentials from.
+http_method:
+  label: HTTP Method
+  help: The HTTP method the library uses when requesting credentials from Vault.
+http_request_body:
+  label: HTTP Request Body
+  help: The body of the HTTP request the library sends to Vault when requesting credentials. Only allowed when http_method is set to POST.
```
Comment (randallmorey) [inline addons/core/translations/form/en-us.yaml:126]: The field name `http_request_body` is probably too generic to mention a specific resource type or cred store type, or particular validation constraints.  I'm not sure how to handle this situation.  Thoughts?

File: addons/core/translations/form/en-us.yaml [hunk 2]
```diff
@@ -115,3 +115,12 @@ value:
 callback_url:
   label: Callback URL
   help: The URL to which the third-party provider sends the authentication response.  Paste this into the provider configuration.
+vault_path:
+  label: Vault Path
+  help: The path in Vault to request credentials from.
+http_method:
+  label: HTTP Method
+  help: The HTTP method the library uses when requesting credentials from Vault.
```
Comment (randallmorey) [inline addons/core/translations/form/en-us.yaml:123]: The field name `http_method` is too generic to mention a specific resource type or cred store type, or particular validation constraints.  I'm not sure how to handle this situation.  Thoughts?

File: addons/core/translations/form/en-us.yaml [hunk 3]
```diff
@@ -115,3 +115,12 @@ value:
 callback_url:
   label: Callback URL
   help: The URL to which the third-party provider sends the authentication response.  Paste this into the provider configuration.
+vault_path:
+  label: Vault Path
+  help: The path in Vault to request credentials from.
```
Comment (randallmorey) [inline addons/core/translations/form/en-us.yaml:120]: This seems fine because of the field name.

Comment (randallmorey) [reply addons/core/translations/form/en-us.yaml:120]: Did this description come from the Go code?  They usually include decent descriptions.

**Review by randallmorey — CHANGES_REQUESTED** [review]

File: ui/admin/app/components/form/credential-library/index.hbs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      </field.errors>
+    {{/if}}
+  </form.textarea>
+
+  <form.input
+    @name='vault_path'
+    @type='text'
+    @value={{@model.attributes.path}}
+    @label={{t 'form.vault_path.label'}}
+    @error={{@model.errors.name}}
```
Comment (randallmorey) [inline ui/admin/app/components/form/credential-library/index.hbs:52]: field name

File: ui/admin/app/components/form/credential-library/index.hbs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    {{#if @model.errors.name}}
+      <field.errors as |errors|>
+        {{#each @model.errors.name as |error|}}
+          <errors.message>{{error.message}}</errors.message>
+        {{/each}}
+      </field.errors>
+    {{/if}}
+  </form.input>
+
+  <form.radioGroup
```
Comment (randallmorey) [inline ui/admin/app/components/form/credential-library/index.hbs:66]: What about errors display?

File: ui/admin/app/components/form/credential-library/index.hbs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      <radioGroup.radio @label={{httpMethod}} @value={{httpMethod}} />
+    {{/each}}
+  </form.radioGroup>
+
+  <form.textarea
+    @name='http_request_body'
+    @type='text'
+    @value={{@model.attributes.http_request_body}}
+    @label={{t 'form.http_request_body.label'}}
+    @error={{@model.errors.description}}
```
Comment (randallmorey) [inline ui/admin/app/components/form/credential-library/index.hbs:82]: `@model.errors.attributes_http_request_body`

File: ui/admin/app/components/form/credential-library/index.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+   * @type {object}
+   */
+  httpMethods = options.http_methods;
+
+  /**
+   * Only allow HTTP request body field if http_method is set to POST.
+   * @type {boolean}
+   */
+  @computed('args.model.attributes.http_method', 'isSaving')
+  get cannotSaveHttpRequestBody() {
```
Comment (randallmorey) [inline ui/admin/app/components/form/credential-library/index.js:17]: This is specifically for the disabled state, not save-ability per se.

File: addons/core/translations/form/en-us.yaml (hunk 1 above)
Comment (gsusmi) [reply addons/core/translations/form/en-us.yaml:126]: Should this be nested to indicate it's vault related?
```
vault:
  path:
  http_method:
  http_body:
```
or for additional structure
```
vault:
  path:
  http:
    method:
    body:
    post_body:
```

**Review by randallmorey — CHANGES_REQUESTED** [review]

File: addons/core/translations/form/en-us.yaml (hunk 1 above)
Comment (randallmorey) [reply addons/core/translations/form/en-us.yaml:126]: I'm sort of thinking the pattern here would be to move it under the resource-specific translations into say a `form` key?

File: addons/core/translations/form/en-us.yaml (hunk 3 above)
Comment (gsusmi) [reply addons/core/translations/form/en-us.yaml:120]: From credential library [docs](https://www.boundaryproject.io/docs/concepts/domain-model/credential-libraries#vault-credential-library-attributes) on boundaryproject.io :).

File: ui/admin/app/components/form/credential-library/index.hbs (hunk 5 above)
Comment (gsusmi) [reply ui/admin/app/components/form/credential-library/index.hbs:66]: Oops. Will update.

File: ui/admin/app/components/form/credential-library/index.js (hunk 7 above)
Comment (gsusmi) [reply ui/admin/app/components/form/credential-library/index.js:17]: Good point. If model save state isn't checked in this, is there another way the radiobutton group can be disabled during save?
Or is your comment referring to the name of the function?

File: addons/api/addon/models/credential-library.js [hunk 8]
```diff
@@ -2,13 +2,24 @@ import GeneratedCredentialLibraryModel from '../generated/models/credential-libr
 import { attr } from '@ember-data/model';
 import { fragment } from 'ember-data-model-fragments/attributes';
 
+/**
+ * Enum options for credential library.
+ */
+export const options = {
+  http_methods: ['GET', 'POST'],
```
[the remark this answers is not in the document]
Comment (gsusmi) [reply addons/api/addon/models/credential-library.js:9]: The `http_method` field is a string. I thought using `http_methods` as array was easiest way to populate it's associated radiobuttons in form. Better ways to define this? I guess I can move this into the form component class.

File: addons/core/translations/form/en-us.yaml (hunk 1 above)
Comment (gsusmi) [reply addons/core/translations/form/en-us.yaml:126]: I like the sound of it. We have tried to keep resource-specific translations pretty flat so how does this look:
```
credential-library:
  title:
  form:
    vault:
      path:
        label:
        description:
      method:
      body:
```

Main reason behind using `vault` tag is coz `method` and `body` descriptions are vault specific.

File: ui/admin/app/routes/scopes/scope/credential-stores/credential-store/credential-libraries/new.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+    this.render(
+      'scopes/scope/credential-stores/credential-store/credential-libraries/new/-header',
+      {
+        into: 'scopes/scope/credential-stores/credential-store',
+        outlet: 'header',
+      }
+    );
+
+    this.render('_empty', {
```
Comment (calcaide) [inline ui/admin/app/routes/scopes/scope/credential-stores/credential-store/credential-libraries/new.js:35]: This is a question more than a comment. I didn't see these before. Is this a way to could define the `renderTemplate` before having the template file ready?

**Comment by calcaide** [comment]
I know this is a draft and you already have it plan, but before moving from draft to ready, a rebase with main might be need since I merged changes recently in some of the `credential-store` routes and templates.

Nice work!! :partying_face:

File: ui/admin/app/routes/scopes/scope/credential-stores/credential-store/credential-libraries/new.js (hunk 9 above)
Comment (gsusmi) [reply ui/admin/app/routes/scopes/scope/credential-stores/credential-store/credential-libraries/new.js:35]: This is going to render an [_empty.hbs](https://github.com/hashicorp/boundary-ui/blob/main/ui/admin/app/templates/_empty.hbs) template file that's available at `templates` root. It was created in a previous iteration to provide us with a way to clear named outlets as needed. Render method does complain loudly when attempting to render a file that doesn't exist.

**Comment by gsusmi** [comment]
credential-library-error-handling

Radiobutton group component doesn't support error handling yet. Will introduce this in a future PR onto Rose addon.

File: addons/core/translations/form/en-us.yaml (hunk 1 above)
Comment (gsusmi) [reply addons/core/translations/form/en-us.yaml]: Moved labels into `resources.credential-library.form`

File: addons/core/translations/form/en-us.yaml (hunk 3 above)
Comment (gsusmi) [reply addons/core/translations/form/en-us.yaml]: Updated description to `The path the library uses when requesting credentials from Vault.`

File: addons/api/addon/models/credential-library.js [hunk 10]
```diff
@@ -2,13 +2,24 @@ import GeneratedCredentialLibraryModel from '../generated/models/credential-libr
...
   // These attributes exist solely to capture errors on nested fields.
   // See the application adapter's error normalization method for
   // more information.
 
-  @attr('string', { readOnly: true }) attributes_http_method;
+  @attr('string', {
+    readOnly: true,
+    defaultValue: options.http_methods[0],
+  })
+  attributes_http_method;
```
[the remark this answers is not in the document]
Comment (gsusmi) [reply addons/api/addon/models/credential-library.js]: Gotcha. I should have read the comment above that describes this. Updated to set default value in attributes fragment.

File: ui/admin/app/components/form/credential-library/index.js (hunk 7 above)
Comment (gsusmi) [reply ui/admin/app/components/form/credential-library/index.js]: Refactored this into `isHttpRequestBodyAllowed` without using `isSaving` method. Reflects design change where text area is only shown when http_method value is POST.

File: addons/api/addon/models/credential-library.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (gsusmi) [reply addons/api/addon/models/credential-library.js]: Renamed to `http_method` to match attribute name.

File: ui/admin/app/routes/scopes/scope/credential-stores/credential-store/credential-libraries/new.js (hunk 9 above)
Comment (calcaide) [reply ui/admin/app/routes/scopes/scope/credential-stores/credential-store/credential-libraries/new.js:35]: Awesome! thanks :)

**Review by calcaide — APPROVED** [review]
LGTM!! Awesome job

**Review by randallmorey — APPROVED** [review]

**Pull request merged**